### PR TITLE
Add TCPC drivers for TCPCI and PS8815

### PR DESCRIPTION
**Pull request by barnas-michal**
Improve USB-C PD stack support for different TCPC drivers.
Add support for TCPCI compliant chips and add driver for Parade PS8815 using TCPC subsystem.

### CODE REVIEW

**Comment by MaureenHelm** [comment]
@sambhurst please take a look

@barnas-michal please rebase

File: drivers/usb_c/tcpc/tcpci.c [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+		}
+	}
+
+	if (ret != 0) {
+		LOG_ERR(LOG_COMM_ERR_STR, i2c->bus->name, i2c->addr, "w16", reg, value);
+	}
+
+	return ret;
+}
+
```
Comment (sambhurst) [inline drivers/usb_c/tcpc/tcpci.c:309]: should there be a tcpci_update_reg16 function?

File: drivers/usb_c/tcpc/tcpci.c [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+enum tcpc_alert tcpci_alert_reg_to_enum(uint16_t reg)
+{
+	if (reg & TCPC_REG_ALERT_RX_HARD_RST) {
```
Comment (sambhurst) [inline drivers/usb_c/tcpc/tcpci.c:312]: Are these checks in order of importance?

File: drivers/usb_c/tcpc/tcpci.c [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+	}
+
+	if (ret != 0) {
+		LOG_ERR(LOG_COMM_ERR_STR, i2c->bus->name, i2c->addr, "w16", reg, value);
+	}
+
+	return ret;
+}
+
+enum tcpc_alert tcpci_alert_reg_to_enum(uint16_t reg)
```
Comment (sambhurst) [inline drivers/usb_c/tcpc/tcpci.c:310]: So this function needs to be repeatedly called to get all the pending alerts. Maybe return 0 to signal that no alerts are pending, instead of -1 to signal an error.

File: drivers/usb_c/vbus/usbc_vbus_tcpci.c [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+#include <zephyr/drivers/usb_c/usbc_pd.h>
+#include <zephyr/usb_c/tcpci.h>
+
+/** Configuration structure for device instances */
+struct vbus_tcpci_cfg {
+	/** I2C bus and address used for communication, set from parent node of device. */
+	const struct i2c_dt_spec i2c;
+};
+
+static int tcpci_measure(const struct device *dev, int *vbus_meas)
```
Comment (sambhurst) [inline drivers/usb_c/vbus/usbc_vbus_tcpci.c:25]: nit: check if vbus_meas is null and return error if it is

File: drivers/usb_c/tcpc/ps8xxx.c [hunk 5]
```diff
@@ -108,6 +110,17 @@ int tcpci_tcpm_get_cc(const struct device *dev, enum tc_cc_voltage_state *cc1,
...
 
+#ifdef CONFIG_USBC_CSM_SINK_ONLY
+	if (!data->cc_changed) {
+		*cc1 = data->cc1;
+		*cc2 = data->cc2;
+
+		return 0;
+	}
+
+	data->cc_changed = 0;
```
Comment (sambhurst) [inline drivers/usb_c/tcpc/ps8xxx.c:121]: data->cc_changed = false

File: drivers/usb_c/tcpc/ps8xxx.c [hunk 6]
```diff
@@ -193,6 +206,8 @@ int ps8xxx_tcpc_select_rp_value(const struct device *dev, enum tc_rp_value rp)
 	const struct ps8xxx_cfg *cfg = dev->config;
 	struct ps8xxx_data *data = dev->data;
 
+	data->cc_changed = 1;
```
Comment (sambhurst) [inline drivers/usb_c/tcpc/ps8xxx.c:209]: data->cc_changed = true;

File: drivers/usb_c/tcpc/ps8xxx.c [hunk 7]
```diff
@@ -218,6 +233,8 @@ int ps8xxx_tcpc_set_cc(const struct device *dev, enum tc_cc_pull pull)
 		return -EIO;
 	}
 
+	data->cc_changed = 1;
```
Comment (sambhurst) [inline drivers/usb_c/tcpc/ps8xxx.c:236]: data->cc_changed = true; Same below

File: drivers/usb_c/tcpc/tcpci.c (hunk 1 above)
Comment (barnas-michal) [reply drivers/usb_c/tcpc/tcpci.c:309]: There's no need to, since it wasn't needed for the driver implemented and I didn't want to write function that won't be used at the moment

File: drivers/usb_c/tcpc/tcpci.c (hunk 2 above)
Comment (barnas-michal) [reply drivers/usb_c/tcpc/tcpci.c:312]: Only the hard reset has precedence, because it by design should be handled resetting everything else, other options are in order of bits in the register.

File: drivers/usb_c/tcpc/tcpci.c (hunk 3 above)
Comment (barnas-michal) [reply drivers/usb_c/tcpc/tcpci.c:310]: I wanted these enums to have the same value as bit position in the alert register, because if we have this enum, we can easily clear the bits in the register with BIT(enum_value), where BIT(enum_value - 1) is still valid, but less obvious. But if you think it is better, I can change that.

File: drivers/usb_c/vbus/usbc_vbus_tcpci.c (hunk 4 above)
Comment (barnas-michal) [reply drivers/usb_c/vbus/usbc_vbus_tcpci.c:25]: I've copied the code from ADC VBUS and used assert for that.

File: drivers/usb_c/tcpc/ps8xxx.c (hunk 6 above)
Comment (barnas-michal) [reply drivers/usb_c/tcpc/ps8xxx.c]: Done

File: drivers/usb_c/tcpc/ps8xxx.c (hunk 7 above)
Comment (barnas-michal) [reply drivers/usb_c/tcpc/ps8xxx.c]: Done

File: drivers/usb_c/tcpc/Kconfig.tcpc_tcpci [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+config USBC_TCPC_TCPCI_I2C_RETRIES
+	int "I2C communication retries"
+	default 2
+	help
+	  Number of I2C transaction tries that will be performed for each request.
+	  Some TCPCs are going into deep sleep mode when no charger is connected and won't respond
+	  to the i2c address immediately. If device won't respond after retries, it means that
+	  it is not responsible or is not connected.
+
+config USBC_TCPC_TCPCI_TRANSMIT_RETRIES
```
Comment (keith-zephyr) [inline drivers/usb_c/tcpc/Kconfig.tcpc_tcpci:25]: Should this be a property of the tcpci driver's devicetree node?

File: drivers/usb_c/tcpc/Kconfig.tcpc_tcpci [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+	int "Data transmit retries"
+	default 2
+	range 0 3
+	depends on USBC_TCPC_TCPCI
+	help
+	  Maximum number of packet retransmissions done by TCPC
```
Comment (keith-zephyr) [inline drivers/usb_c/tcpc/Kconfig.tcpc_tcpci:31]: Regardless of whether this is a Kconfig or devicetree property, it would be helpful to note this setting is written to the "Retry Counter" of the TRANSMIT register of the TCPCI.

File: drivers/usb_c/tcpc/Kconfig.tcpc_tcpci [hunk 10]
```diff
@@ -31,3 +31,27 @@ config USBC_TCPC_TCPCI_TRANSMIT_RETRIES
...
+	help
+	  Delay between each try of the TCPC initialization
+
+config USBC_TCPC_PS8XXX_INIT_RETRIES
+	int "PS8xxx init retries"
+	default 10
+	help
+	  Number of initialization tries that will be performed
+
+endif
```
Comment (keith-zephyr) [inline drivers/usb_c/tcpc/Kconfig.tcpc_tcpci:57]: Each driver should have it's own Kconfig file.

File: drivers/usb_c/tcpc/ps8xxx.c [hunk 11]
```diff
@@ -0,0 +1,799 @@
...
+		}
+
+		return -EAGAIN;
+	}
+
+	LOG_INF("PS8xxx TCPC initialized");
+	return 0;
+}
+
+int tcpci_tcpm_get_cc(const struct device *dev, enum tc_cc_voltage_state *cc1,
```
Comment (keith-zephyr) [inline drivers/usb_c/tcpc/ps8xxx.c:99]: This should be moved to tcpci.c.

**Review by keith-zephyr — COMMENTED** [review]
Please also run clang-format on the new files.

File: drivers/usb_c/tcpc/tcpci.c (hunk 2 above)
Comment (barnas-michal) [reply drivers/usb_c/tcpc/tcpci.c:312]: Good idea, done.

File: drivers/usb_c/tcpc/Kconfig.tcpc_tcpci (hunk 8 above)
Comment (barnas-michal) [reply drivers/usb_c/tcpc/Kconfig.tcpc_tcpci:25]: Sounds reasonable, done.

File: drivers/usb_c/tcpc/Kconfig.tcpc_tcpci (hunk 9 above)
Comment (barnas-michal) [reply drivers/usb_c/tcpc/Kconfig.tcpc_tcpci:31]: Right, done.

File: drivers/usb_c/tcpc/Kconfig.tcpc_tcpci (hunk 10 above)
Comment (barnas-michal) [reply drivers/usb_c/tcpc/Kconfig.tcpc_tcpci:57]: Done.

**Review by sambhurst — DISMISSED** [review]

File: drivers/usb_c/tcpc/ps8xxx.c [hunk 12]
```diff
@@ -581,6 +598,7 @@ void ps8xxx_alert_work_cb(struct k_work *work)
 		return;
 	}
 
+	data->cc_changed = true;
```
Comment (keith-zephyr) [inline drivers/usb_c/tcpc/ps8xxx.c:601]: Why is this needed?  This flags a CC change on any interrupt.  If that's desired, I don't think you need to set `cc_changed=true` below in the while loop.

Comment (barnas-michal) [reply drivers/usb_c/tcpc/ps8xxx.c]: I forgot to remove this after inserting the same in specific cases. Fixed.

**Review by keith-zephyr — APPROVED** [review]

**Review by sambhurst — APPROVED** [review]
